exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java

```diff
@@ -424,26 +424,35 @@ public void eval() {
     public static class AgeTimeStampFunction implements DrillSimpleFunc {
         @Param TimeStampHolder left;
         @Param TimeStampHolder right;
+        @Workspace org.joda.time.PeriodType periodType;
         @Output IntervalHolder out;
 
         @Override
         public void setup() {
+            periodType = org.joda.time.PeriodType.forFields(
+                new org.joda.time.DurationFieldType[] {
+                    org.joda.time.DurationFieldType.months(),
+                    org.joda.time.DurationFieldType.days(),
+                    org.joda.time.DurationFieldType.millis()
+                }
+            );
         }
 
         @Override
         public void eval() {
-            long diff = left.value - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            org.joda.time.Period interval = new org.joda.time.Period(right.value, left.value, periodType);
```

Comment:
Using Joda is certainly a quick fix! What was wrong with the `DateUtilities` code? Since that code is used in multiple places, could we fix that code? Or, if the fix is complicated (everything with dates is complicated), should we change all other uses of the same constants and delete the constants in `DateUtilities`?

Also, Joda is long obsolete. Is there a "modern" way to do these calculations?

Reply:
There is no class in Java Time API that would allow working simultaneously with months and hours / minutes / seconds. However, I can try to rewrite the fix using java.time.Period and java.time.Duration.